File: workbook/ch08/data/fibheap.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct Node {
    int key, degree;
    struct Node *parent, *child, *next, *prev;
    int mark;
} Node;

typedef struct FibonacciHeap {
    Node *min;
    int total_nodes;
} FibonacciHeap;

Node* create_node(int key) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->key = key;
    node->degree = 0;
    node->parent = node->child = NULL;
    node->mark = 0;
    node->next = node->prev = node;
    return node;
}

FibonacciHeap* create_heap() {
    FibonacciHeap* heap = (FibonacciHeap*)malloc(sizeof(FibonacciHeap));
    heap->min = NULL;
    heap->total_nodes = 0;
    return heap;
}

Node* merge_lists(Node* a, Node* b) {
    if (!a) return b;
    if (!b) return a;
    Node* temp = a->next;
    a->next = b->next;
    b->next->prev = a;
    b->next = temp;
    temp->prev = b;
    return (a->key < b->key) ? a : b;
}

void insert(FibonacciHeap* heap, int key) {
    Node* node = create_node(key);
    heap->min = merge_lists(heap->min, node);
    heap->total_nodes++;
}

void remove_from_list(Node* node) {
    if (node->next == node) return;
    node->prev->next = node->next;
    node->next->prev = node->prev;
}

void link(Node* child, Node* parent) {
    remove_from_list(child);
    child->next = child->prev = child;
    parent->child = merge_lists(parent->child, child);
    child->parent = parent;
    parent->degree++;
    child->mark = 0;
}

void consolidate(FibonacciHeap* heap) {
    int max_degree = (int)(log(heap->total_nodes) / log(2)) + 1;
    Node* degree_table[max_degree];
    for (int i = 0; i < max_degree; i++) degree_table[i] = NULL;
    
    Node* start = heap->min;
    Node* node = start;
    do {
        Node* x = node;
        node = node->next;
        int d = x->degree;
        while (degree_table[d]) {
            Node* y = degree_table[d];
            if (x->key > y->key) {
                Node* temp = x;
                x = y;
                y = temp;
            }
            link(y, x);
            degree_table[d] = NULL;
            d++;
        }
        degree_table[d] = x;
    } while (node != start);
    
    heap->min = NULL;
    for (int i = 0; i < max_degree; i++) {
        if (degree_table[i]) {
            heap->min = merge_lists(heap->min, degree_table[i]);
        }
    }
}

Node* extract_min(FibonacciHeap* heap) {
    Node* min_node = heap->min;
    if (min_node) {
        if (min_node->child) {
            Node* child = min_node->child;
            do {
                child->parent = NULL;
                child = child->next;
            } while (child != min_node->child);
            heap->min = merge_lists(heap->min, min_node->child);
        }
        remove_from_list(min_node);
        if (min_node == min_node->next) {
            heap->min = NULL;
        } else {
            heap->min = min_node->next;
            consolidate(heap);
        }
        heap->total_nodes--;
    }
    return min_node;
}


int main() {
    FibonacciHeap* heap = create_heap();
    insert(heap, 10);
    insert(heap, 3);
    insert(heap, 7);
    printf("Extracted min: %d\n", extract_min(heap)->key);  // Output: 3
    return 0;
}

File: workbook/ch07/data/common/fibheap.py
import math

class Node:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.mark = False
        self.next = self
        self.prev = self

class FibonacciHeap:
    def __init__(self):
        self.min = None
        self.total_nodes = 0

    def insert(self, key):
        node = Node(key)
        self.min = self._merge_lists(self.min, node)
        self.total_nodes += 1
        return node

    def _merge_lists(self, a, b):
        if not a: return b
        if not b: return a
        
        a.next, b.prev, a.prev, b.next = b, a, b.next, a.prev
        return a if a.key < b.key else b

    def extract_min(self):
        min_node = self.min
        if min_node:
            if min_node.child:
                children = []
                child = min_node.child
                while True:
                    children.append(child)
                    child = child.next
                    if child == min_node.child:
                        break
                
                for child in children:
                    child.parent = None
                    self.min = self._merge_lists(self.min, child)
            
            self._remove_from_list(min_node)
            if min_node == min_node.next:
                self.min = None
            else:
                self.min = min_node.next
                self._consolidate()
            
            self.total_nodes -= 1
        return min_node

    def _remove_from_list(self, node):
        if node.next == node:
            return
        node.prev.next = node.next
        node.next.prev = node.prev

    def _consolidate(self):
        max_degree = int(math.log(self.total_nodes) * 2) if self.total_nodes > 0 else 1
        degree_table = [None] * max_degree

        nodes = []
        x = self.min
        if x:
            while True:
                nodes.append(x)
                x = x.next
                if x == self.min:
                    break
        
        for node in nodes:
            d = node.degree
            while degree_table[d]:
                other = degree_table[d]
                if node.key > other.key:
                    node, other = other, node
                self._link(other, node)
                degree_table[d] = None
                d += 1
            degree_table[d] = node
        
        self.min = None
        for entry in degree_table:
            if entry:
                self.min = self._merge_lists(self.min, entry)

    def _link(self, child, parent):
        self._remove_from_list(child)
        child.next = child.prev = child
        parent.child = self._merge_lists(parent.child, child)
        child.parent = parent
        parent.degree += 1
        child.mark = False

fib_heap = FibonacciHeap()
fib_heap.insert(10)
fib_heap.insert(3)
fib_heap.insert(7)
print("Extracted min:", fib_heap.extract_min().key)  # Output: 3
